### PR TITLE
Handle Transformer Engine delayed gradients in MFSDP

### PR DESCRIPTION
**Pull request by wujingyue**
## Summary

MFSDP normally treats PyTorch post-accumulate-grad hooks as the gradient-ready signal. Transformer Engine delayed-wgrad parameters instead materialize their gradients in `backward_dw()`.

Register MFSDP’s existing gradient-completion hook through TE’s `register_wgrad_accumulation_and_reduce_hooks()` for parameters marked with TE’s `skip_backward_post_hook` contract. Ordinary parameters retain the normal PyTorch hook.

Add a direct, single-rank TE `Linear` test that verifies MFSDP remains in backward after autograd and completes only after `backward_dw()`.

## Dependency

Depends on #6695, which shares the MFSDP gradient-completion hook used here. This branch is based on that PR’s head.

## Tests

- `python -m torch.distributed.run --nproc-per-node 1 -m pytest -q --capture=fd tests/unit_tests/distributed/mfsdp_v2/test_fully_shard.py::test_fully_shard_waits_for_delayed_te_weight_gradient`
- `ruff check megatron/core/distributed/fsdp/src/megatron_fsdp/experimental/module.py`

### CODE REVIEW

File: megatron/core/distributed/fsdp/src/megatron_fsdp/experimental/module.py [hunk 1]
```diff
@@ -270,24 +255,32 @@ def _register_hooks(self) -> None:
...
+                continue
+            for fsdp_parameter in group.fsdp_parameters:
+                parameter = fsdp_parameter.unsharded
+                # ``skip_backward_post_hook`` is TE's delayed-wgrad contract: these
+                # gradients are materialized by ``backward_dw()``, not autograd.
+                if not getattr(parameter, "skip_backward_post_hook", False):
+                    parameter.register_post_accumulate_grad_hook(grad_hook)
+                    continue
+
+                module_fqn, _, _ = fsdp_parameter.fqns[0].rpartition(".")
```
Comment (wujingyue) [inline megatron/core/distributed/fsdp/src/megatron_fsdp/experimental/module.py]: This looks ugly. I think we should iterate through the submodules instead of iterating through the parameters and finding their owning modules. Also, not sure if there's a bug with tied weights -- need to double check.

File: tests/unit_tests/distributed/mfsdp_v2/test_fully_shard.py [hunk 2]
```diff
@@ -248,6 +249,63 @@ def train(model, optimizer, log_prefix) -> list[torch.Tensor]:
...
+    x = torch.randn(4, 16, device=device, dtype=torch.bfloat16, requires_grad=True)
+    model(x).float().square().mean().backward()
+    assert model.weight.grad is None
+    assert model.phase is FsdpModule.Phase.BACKWARD
+
+    model.backward_dw()
+    torch.cuda.synchronize(device)
+
+    assert model.weight.grad is not None
+    assert model.phase is FsdpModule.Phase.RESTING
```
Comment (shjwudp) [inline tests/unit_tests/distributed/mfsdp_v2/test_fully_shard.py:278]: It does not seem to proof the `post_backward` is specifically triggered by `backward_dw`. Would it be better to register distinct mock hooks for `register_post_accumulate_grad_hook` and `register_wgrad_accumulation_and_reduce_hooks`, then assert their respective invocations?

Comment (wujingyue) [reply tests/unit_tests/distributed/mfsdp_v2/test_fully_shard.py:278]: https://testing.googleblog.com/2013/08/testing-on-toilet-test-behavior-not.html

I'm pretty comfortable with the test as is. It verifies phase==BACKWARD right after `backward()` **and** phase==RESTING right after `backward_dw()`. Mock is sometimes useful but often makes the test focused on implementation not behavior.

**Review by shjwudp — APPROVED** [review]

File: tests/unit_tests/distributed/mfsdp_v2/test_fully_shard.py (hunk 2 above)
Comment (shjwudp) [reply tests/unit_tests/distributed/mfsdp_v2/test_fully_shard.py:278]: Thanks! No problem.

**Comment by wujingyue** [comment]
Hey @rapatel, have you ever needed to delay_grad an **FsdpModule**? This PR is probably unnecessary unless you need it. For MoE 1f1b, I'm seeing the **submodules** of an FsdpModule get `delay_grad`ed, not the FsdpModule itself.

**Comment by rapatel** [comment]
@wujingyue does "delay_grad an FsdpModule" mean that we would delay every wgrad calculation of layer[n] to layer[n-1]? Not sure what that buys us, sounds like it would just shift every wgrad by one layer.

**Comment by wujingyue** [comment]
> delay every wgrad calculation of layer[n] to layer[n-1]

Not necessarily. We could `fully_shard` a **submodule** of `layer[n]`. For example, we might fully_shard the MLP layer in a TransformerLayer, and delay/split the MLP's wgrad to overlap with communication. However, in practice, I guess we just `fully_shard` the TransformerLayer itself and delay_wgrad non-FsdpModules. 

image
